hipcheck/src/plugin/parser/plugin_manifest.rs

```diff
@@ -0,0 +1,418 @@
+pub struct Name(pub String);
+string_newtype_parse_kdl_node!(Name, "name");
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Version(pub String);
+string_newtype_parse_kdl_node!(Version, "version");
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct License(pub String);
+string_newtype_parse_kdl_node!(License, "license");
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+// TODO: target-triple enum
+pub struct Entrypoints(pub HashMap<String, String>);
+
+impl Entrypoints {
+	pub fn new() -> Self {
+		Self(HashMap::new())
+	}
+
+	pub fn insert(&mut self, arch: String, entrypoint: String) -> Result<(), Error> {
+		match self.0.insert(arch.clone(), entrypoint) {
+			Some(_duplicate_key) => Err(hc_error!("Multiple entrypoints specified for {}", arch)),
+			None => Ok(()),
+		}
+	}
+
+	pub fn iter(&self) -> impl Iterator<Item = (&String, &String)> {
+		self.0.iter()
+	}
+}
+
+impl ParseKdlNode for Entrypoints {
+	fn kdl_key() -> &'static str {
+		"entrypoint"
+	}
+
+	fn parse_node(node: &KdlNode) -> Option<Self> {
+		if node.name().to_string().as_str() != Self::kdl_key() {
+			return None;
+		}
+		let mut entrypoints = Entrypoints::new();
+		for entrypoint_spec in node.children()?.nodes() {
+			// per RFD #0004, the value for "arch" is of type String
+			let arch = entrypoint_spec
+				.get("arch")?
+				.value()
+				.as_string()?
+				.to_string();
+			// per RFD #0004, the actual entrypoint is the first positional arg after "arch" and is
+			// of type String
+			let entrypoint = entrypoint_spec
+				.entries()
+				.get(1)?
+				.value()
+				.as_string()?
+				.to_string();
+			if let Err(e) = entrypoints.insert(arch.clone(), entrypoint) {
+				log::error!("Duplicate entrypoint detected for [{arch}]");
+				return None;
+			}
+		}
+		Some(entrypoints)
+	}
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PluginDependency {
+	pub name: String,
+	pub version: String,
+	// NOTE: until Hipcheck supports a registry, this is effectively required
+	pub manifest: Option<String>,
+}
+
+impl PluginDependency {
+	pub fn new(name: String, version: String, manifest: Option<String>) -> Self {
+		Self {
+			name,
+			version,
+			manifest,
+		}
+	}
+}
+
+impl ParseKdlNode for PluginDependency {
+	fn kdl_key() -> &'static str {
+		"plugin"
+	}
+
+	fn parse_node(node: &KdlNode) -> Option<Self> {
+		if node.name().to_string().as_str() != Self::kdl_key() {
+			return None;
+		}
+
+		// per RFD #0004, the name is the first positional entry and has type String
+		let name = node.entries().first()?.value().as_string()?.to_string();
+		let version = node.get("version")?.value().as_string()?.to_string();
+		let manifest = match node.get("manifest") {
+			Some(manifest) => Some(manifest.value().as_string()?.to_string()),
+			None => None,
+		};
+
+		Some(Self {
+			name,
+			version,
+			manifest,
+		})
+	}
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PluginDependencyList(Vec<PluginDependency>);
+
+impl PluginDependencyList {
```

Comment:
Is there a reason not to implement this as `type PluginDependencyList = Vec<PluginDependency>;`? AFAIK you can `impl` a trait we define on a type we don't define.

Reply:
I was thinking that it would be easier to add custom methods for dealing with all of the dependencies

Reply:
I'm personally not a fan of the shallow-wrapper design, especially because it makes it so we have to re-define any `Vec` function we want to use on a `PluginDependencyList`. I'd advocate for just changing the internal `Vec` to `pub` so it can be accessed directly if need be.

Reply:
I exposed the wrapped Vec with a `pub`. I like not needing to redefine all of the functionality of `Vec` by just using `pub`